sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java

```diff
@@ -272,52 +337,320 @@ public Read withUpToDateThreshold(Duration upToDateThreshold) {
+       * producer properties.
+       *
+       * <p>More details can be found here:
+       * <a href="https://docs.aws.amazon.com/streams/latest/dev/kinesis-kpl-concepts.html">KPL Key Concepts</a> and
+       * <a href="https://docs.aws.amazon.com/streams/latest/dev/kinesis-kpl-config.html">Configuring the KPL</a>
+       */
+      @ProcessElement
+      public void processElement(ProcessContext c) throws Exception {
+        checkForFailures();
+
+        // Need to avoid keeping too many futures in producer's map to prevent OOM.
+        // In usual case, it should exit immediately.
+        flush(MAX_NUM_RECORDS);
+
+        ByteBuffer data = ByteBuffer.wrap(c.element());
+        String partitionKey = spec.getPartitionKey();
+        String explicitHashKey = null;
+
+        // Use custom partitioner
+        if (partitioner != null) {
+          partitionKey = partitioner.getPartitionKey(c.element());
+          explicitHashKey = partitioner.getExplicitHashKey(c.element());
         }
-        return clientBuilder.build();
+
+        ListenableFuture<UserRecordResult> f =
+            producer.addUserRecord(spec.getStreamName(), partitionKey, explicitHashKey, data);
+        Futures.addCallback(f, new UserRecordResultFutureCallback());
       }
 
-      @Override
-      public AmazonCloudWatch getCloudWatchClient() {
-        AmazonCloudWatchClientBuilder clientBuilder =
-            AmazonCloudWatchClientBuilder.standard().withCredentials(getCredentialsProvider());
-        if (serviceEndpoint == null) {
-          clientBuilder.withRegion(region);
-        } else {
-          clientBuilder.withEndpointConfiguration(
-              new AwsClientBuilder.EndpointConfiguration(serviceEndpoint, region.getName()));
+      @FinishBundle
+      public void finishBundle() throws Exception {
+        checkForFailures();
+
+        // Flush all outstanding records, blocking call
+        flushAll();
+      }
+
+      @Teardown
+      public void tearDown() throws Exception {
+        if (producer != null) {
+          producer.destroy();
+          producer = null;
+        }
+      }
+
+      /**
+       * Flush outstanding records until the total number will be less than required or
+       * the number of retries will be exhausted. The retry timeout starts from 1 second and it
+       * doubles on every iteration.
+       */
+      private void flush(int numMax) throws InterruptedException, IOException {
+        int retries = spec.getRetries();
+        int numOutstandingRecords = producer.getOutstandingRecordsCount();
+        int retryTimeout = 1000; // initial timeout, 1 sec
+
+        while (numOutstandingRecords > numMax && retries-- > 0) {
+          producer.flush();
+          // wait until outstanding records will be flushed
+          Thread.sleep(retryTimeout);
+          numOutstandingRecords = producer.getOutstandingRecordsCount();
+          retryTimeout *= 2; // exponential backoff
```

Comment:
Seems that with the default number of 10 retries, we'll wait for up to 2047 seconds (1 + 2 + .. + 1024) ? Fine with me, just wanted to check that it's intentional.

Reply:
Actually, in my last commit (a43afb9f3a09086f20ace8471b9717a031f733f1), I reduced default number of retries to 6, so the total time of retry is about 1 minute. I believe it should be fine.